trin-core/src/utp/stream.rs

```diff
@@ -766,35 +844,337 @@ impl UtpSocket {
         debug!("self.rtt: {}", self.rtt);
         debug!("self.congestion_timeout: {}", self.congestion_timeout);
     }
+}
 
-    fn handle_finalize_packet(&mut self) {
-        if self.state == SocketState::Connected {
-            self.state = SocketState::Disconnected;
-
-            let mut packet_reply = Packet::new();
-            packet_reply.set_type(PacketType::State);
-            packet_reply.set_connection_id(self.sender_connection_id);
-            // TODO: add timestamp difference when we set the delay to self.delay field
-            packet_reply.set_seq_nr(self.seq_nr);
-            packet_reply.set_ack_nr(self.ack_nr);
+#[cfg(test)]
+mod tests {
+    use crate::portalnet::discovery::Discovery;
+    use crate::portalnet::types::messages::PortalnetConfig;
+    use crate::portalnet::Enr;
+    use crate::utp::packets::{Packet, PacketType};
+    use crate::utp::stream::{SocketState, UtpSocket, BUF_SIZE};
+    use discv5::enr::{CombinedKey, EnrBuilder};
+    use rand::Rng;
+    use std::net::Ipv4Addr;
+    use std::sync::Arc;
+
+    fn next_test_port() -> u16 {
+        use std::sync::atomic::{AtomicUsize, Ordering};
+        static NEXT_OFFSET: AtomicUsize = AtomicUsize::new(0);
+        const BASE_PORT: u16 = 9600;
+        BASE_PORT + NEXT_OFFSET.fetch_add(1, Ordering::Relaxed) as u16
+    }
 
-            self.send_packet(&mut packet_reply);
+    fn create_portal_config() -> PortalnetConfig {
+        PortalnetConfig {
+            listen_port: next_test_port(),
+            internal_ip: true,
+            ..Default::default()
         }
     }
 
-    fn handle_syn_packet(&mut self, packet: Packet) {
-        self.receiver_connection_id = packet.connection_id() + 1;
-        self.sender_connection_id = packet.connection_id();
-        self.seq_nr = rand();
-        self.ack_nr = packet.seq_nr();
-        self.state = SocketState::SynRecv;
+    fn generate_random_enr() -> Enr {
```

Comment:
Is it worth pulling this out as a util fn and putting it into `trin-core/src/utils/node_id.rs`? Could you see it being useful in other places in the codebase?

Reply:
I put it in `node_id.rs` for now with  `#[cfg(test)]` flag, it is already used in `overlay_service`.